cosign/verifier.go

```diff
@@ -0,0 +1,324 @@
+	Name string
+
+	// VerifyCommand is a stuct contains params that executes the verification process.
+	// TODO: Update this field to a struct hold multi-VerfiyCommands
+	*verify.VerifyCommand
+}
+
+// Verifier is a ratify.Verifier implementation that verifies cosign
+// signatures.
+type Verifier struct {
+	name       string
+	truststore TrustStore
+	*cosign.CheckOpts
+}
+
+// NewVerifier creates a new cosign verifier.
+func NewVerifier(opts *VerifierOptions) (*Verifier, error) {
+	checkOpts, err := NewCheckOpts(context.Background(), opts.VerifyCommand)
+	if err != nil {
+		return nil, fmt.Errorf("failed to update signature verifier keys: %w", err)
+	}
+
+	return &Verifier{
+		// Set the name of the verifier from the provided options.
+		name: opts.Name,
+		// Set the keys map from the provided options.
+		truststore: NewWithOpts(opts),
+		// Set the cosign check options from the provided options.
+		CheckOpts: checkOpts,
+	}, nil
+}
+
+// NewCheckOpts updates the signature verifierOpts by verifierOptions.
+func NewCheckOpts(ctx context.Context, c *verify.VerifyCommand) (opts *cosign.CheckOpts, err error) {
+	// initialize the cosign check options
+	opts = &cosign.CheckOpts{}
+
+	if c.CheckClaims {
+		opts.ClaimVerifier = cosign.SimpleClaimVerifier
+	}
+
+	// If we are using signed timestamps, we need to load the TSA certificates
+	if c.TSACertChainPath != "" || c.UseSignedTimestamps {
+		tsaCertificates, err := cosign.GetTSACerts(ctx, c.TSACertChainPath, cosign.GetTufTargets)
+		if err != nil {
+			return nil, fmt.Errorf("unable to load TSA certificates: %w", err)
+		}
+		opts.TSACertificate = tsaCertificates.LeafCert
+		opts.TSARootCertificates = tsaCertificates.RootCert
+		opts.TSAIntermediateCertificates = tsaCertificates.IntermediateCerts
+	}
+
+	if !c.IgnoreTlog {
+		if c.RekorURL == "" {
+			c.RekorURL = defaultRekorURL
+		}
+
+		rekorClient, err := rekor.NewClient(c.RekorURL)
+		if err != nil {
+			return nil, fmt.Errorf("creating Rekor client: %w", err)
+		}
+		opts.RekorClient = rekorClient
+
+		// This performs an online fetch of the Rekor public keys, but this is needed
+		// for verifying tlog entries (both online and offline).
+		opts.RekorPubKeys, err = cosign.GetRekorPubs(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("getting Rekor public keys: %w", err)
+		}
+	}
+	return opts, nil
+}
+
+// Name returns the name of the verifier.
+func (v *Verifier) Name() string {
+	return v.name
+}
+
+// Type returns the type of the verifier which is always `cosign`.
+func (v *Verifier) Type() string {
+	return cosignVerifierType
+}
+
+// Verifiable returns true if the artifact is a cosign signature.
+func (v *Verifier) Verifiable(artifact ocispec.Descriptor) bool {
+	return artifact.ArtifactType == artifactTypeCosign &&
+		artifact.MediaType == ocispec.MediaTypeImageManifest
+}
+
+// Verify verifies the cosign signature.
+func (v *Verifier) Verify(ctx context.Context, opts *ratify.VerifyOptions) (*ratify.VerificationResult, error) {
```

Comment:
you might need to use the latest version of ratify-go, the VerifyOptions was updated recently.

Reply:
Sure.